File: shibuyacrossing/submissions/accepted/lukas.cpp
//Solution by lukasP (Lukáš Poláček)
#include <iostream>
#include <vector>
#include <string>
#include <set>
#include <algorithm>

#include <cstdio>
#include <cstring>
#include <cmath>
#include <cctype>
using namespace std;

#define rep(i,a,b) for(__typeof(b) i=a; i<(b); ++i)

typedef vector<int> vi;
int main()
{
    int n, m; scanf("%d %d", &n, &m);

    vi pos(n);
    rep(i,0,n) pos[i] = i;

    rep(i,0,m)
    {
        int a, b;
        scanf("%d %d", &a, &b);
        a--; b--;

        if (a > b)
            swap(a, b);
        pos[a]++;
        pos[b]--;
    }

    vi best(n, 1);
    rep(i,1,n)
        rep(j,0,i) if (pos[j] > pos[i])
            best[i] = max(best[i], best[j] + 1);

    cout << *max_element(best.begin(), best.end()) << endl;
}

File: shibuyacrossing/submissions/accepted/lukas.cs
//Solution by lukasP (Lukáš Poláček)
using System;
using System.IO;
using System.Collections.Generic;
using System.Linq;

class Problem
{
    public static void Main ()
    {
        var bs = new BufferedStream(Console.OpenStandardInput());
        StreamReader reader = new StreamReader(bs);

        string[] ss = reader.ReadLine().Split();
        int n = int.Parse(ss[0]), m = int.Parse(ss[1]);

        List<int> pos = Enumerable.Range(0, n).ToList();

        for (int i = 0; i < m; i++)
        {
            ss = reader.ReadLine().Split();
            int a = int.Parse(ss[0]), b = int.Parse(ss[1]);
            if (a > b)
            {
                int tmp = a;
                a = b;
                b = tmp;
            }
            a--; b--;

            pos[a]++;
            pos[b]--;
        }

        List<int> best = Enumerable.Repeat(1, n).ToList();
        for (int i = 0; i < n; i++)
            for (int j = 0; j < i; j++) if (pos[j] > pos[i])
                best[i] = Math.Max(best[i], best[j] + 1);

        Console.WriteLine(best.Max());
    }
}
